FilePathDisk.newDirectoryStream() may fail on remote drive on Windows due to AccessDeniedException in Path.toRealPath()

2022-04-19 03:51:18 database: opening S:/myfolders/var/global/wx/h2Stores_2/nt/nt-h2-store
org.h2.message.DbException: IO Exception: "java.nio.file.AccessDeniedException: S:\\myfolders\\var\\global\\wx\\h2Stores_2\\nt"; "S:/myfolders/var/global/wx/h2Stores_2/nt" [90031-210]
	at org.h2.message.DbException.get(DbException.java:216)
	at org.h2.message.DbException.convertIOException(DbException.java:461)
	at org.h2.store.fs.disk.FilePathDisk.newDirectoryStream(FilePathDisk.java:219)
	at org.h2.store.fs.FileUtils.newDirectoryStream(FileUtils.java:195)
	at org.h2.engine.Database.deleteOldTempFiles(Database.java:1547)
	at org.h2.engine.Database.<init>(Database.java:320)
	at org.h2.engine.Engine.openSession(Engine.java:92)
	at org.h2.engine.Engine.openSession(Engine.java:222)
	at org.h2.engine.Engine.createSession(Engine.java:201)
	at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:338)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:122)
	at org.h2.jdbcx.JdbcDataSource.getXAConnection(JdbcDataSource.java:322)
	at org.h2.jdbcx.JdbcDataSource.getPooledConnection(JdbcDataSource.java:352)
	at org.h2.jdbcx.JdbcConnectionPool.getConnectionNow(JdbcConnectionPool.java:233)
	at org.h2.jdbcx.JdbcConnectionPool.getConnection(JdbcConnectionPool.java:198)
	at com.netbackup.h2.H2JDBCConnectionPool.getConnection(H2JDBCConnectionPool.java:54)
	at org.hibernate.engine.jdbc.connections.internal.DatasourceConnectionProviderImpl.getConnection(DatasourceConnectionProviderImpl.java:122)
	at org.hibernate.internal.NonContextualJdbcConnectionAccess.obtainConnection(NonContextualJdbcConnectionAccess.java:38)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:108)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.getPhysicalConnection(LogicalConnectionManagedImpl.java:138)
	at org.hibernate.internal.SessionImpl.connection(SessionImpl.java:480)
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.beginTransaction(HibernateJpaDialect.java:152)
	at org.springframework.orm.jpa.JpaTransactionManager.doBegin(JpaTransactionManager.java:421)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.startTransaction(AbstractPlatformTransactionManager.java:400)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:373)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.createTransactionIfNecessary(TransactionAspectSupport.java:595)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:382)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:137)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:174)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
	at com.sun.proxy.$Proxy365.count(Unknown Source)
	at com.netbackup.eventlog.service.EventLogServiceImpl.getntCount(EventLogServiceImpl.java:245)
	at sun.reflect.GeneratedMethodAccessor2919.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:388)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
	at com.sun.proxy.$Proxy367.getntCount(Unknown Source)
	at com.netbackup.eventlog.service.EventLogCleanupService.startCleanup(EventLogCleanupService.java:44)
	at com.netbackup.eventlog.config.EventLogCleanupTask.lambda$runCleanupTask$1(EventLogCleanupTask.java:87)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:95)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: org.h2.jdbc.JdbcSQLNonTransientException: IO Exception: "java.nio.file.AccessDeniedException: S:\\myfolders\\var\\global\\wx\\h2Stores_2\\nt"; "S:/myfolders/var/global/wx/h2Stores_2/nt" [90031-210]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:573)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:496)
	... 61 more
Caused by: java.nio.file.AccessDeniedException: S:\myfolders\var\global\wx\h2Stores_2\nt
	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:83)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:90)
	at sun.nio.fs.WindowsLinkSupport.getRealPath(WindowsLinkSupport.java:259)
	at sun.nio.fs.WindowsPath.toRealPath(WindowsPath.java:836)
	at sun.nio.fs.WindowsPath.toRealPath(WindowsPath.java:44)
	at org.h2.store.fs.disk.FilePathDisk.newDirectoryStream(FilePathDisk.java:214)

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,7 +21,11 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
-<li>Nothing yet ...
+<li>Issue #3493: org.h2.tools.DeleteDbFiles won't delete files under certain circumstances
+</li>
+<li>Issue #3486: FilePathDisk.newDirectoryStream() may fail on remote drive on Windows due to AccessDeniedException in Path.toRealPath()
+</li>
+<li>Issue #3484: LOB issue
 </li>
 </ul>
 

--- a/h2/src/main/org/h2/store/FileLister.java
+++ b/h2/src/main/org/h2/store/FileLister.java
@@ -86,10 +86,10 @@ public class FileLister {
     public static ArrayList<String> getDatabaseFiles(String dir, String db,
             boolean all) {
         ArrayList<String> files = new ArrayList<>();
-        // for Windows, File.getCanonicalPath("...b.") returns just "...b"
-        String start = db == null ? null : (FileUtils.toRealPath(dir + "/" + db) + ".");
-        for (String f : FileUtils.newDirectoryStream(dir)) {
+        String start = db == null ? null : db + '.';
+        for (FilePath path : FilePath.get(dir).newDirectoryStream()) {
             boolean ok = false;
+            String f = path.toString();
             if (f.endsWith(Constants.SUFFIX_MV_FILE)) {
                 ok = true;
             } else if (all) {
@@ -102,7 +102,7 @@ public class FileLister {
                 }
             }
             if (ok) {
-                if (db == null || f.startsWith(start)) {
+                if (db == null || path.getName().startsWith(start)) {
                     files.add(f);
                 }
             }

--- a/h2/src/main/org/h2/store/fs/disk/FilePathDisk.java
+++ b/h2/src/main/org/h2/store/fs/disk/FilePathDisk.java
@@ -211,7 +211,7 @@ public class FilePathDisk extends FilePath {
 
     @Override
     public List<FilePath> newDirectoryStream() {
-        try (Stream<Path> files = Files.list(Paths.get(name).toRealPath())) {
+        try (Stream<Path> files = Files.list(toRealPath(Paths.get(name)))) {
             return files.collect(ArrayList::new, (t, u) -> t.add(getPath(u.toString())), ArrayList::addAll);
         } catch (NoSuchFileException e) {
             return Collections.emptyList();
@@ -267,19 +267,27 @@ public class FilePathDisk extends FilePath {
 
     @Override
     public FilePathDisk toRealPath() {
-        Path path = Paths.get(name);
+        return getPath(toRealPath(Paths.get(name)).toString());
+    }
+
+    private static Path toRealPath(Path path) {
         try {
-            return getPath(path.toRealPath().toString());
+            path = path.toRealPath();
         } catch (IOException e) {
             /*
              * File does not exist or isn't accessible, try to get the real path
              * of parent directory.
+             *
+             * toRealPath() can also throw AccessDeniedException on accessible
+             * remote directory on Windows if other directories on remote drive
+             * aren't accessible, but toAbsolutePath() should work.
              */
-            return getPath(toRealPath(path.toAbsolutePath().normalize()).toString());
+            path = parentToRealPath(path.toAbsolutePath().normalize());
         }
+        return path;
     }
 
-    private static Path toRealPath(Path path) {
+    private static Path parentToRealPath(Path path) {
         Path parent = path.getParent();
         if (parent == null) {
             return path;
@@ -287,7 +295,7 @@ public class FilePathDisk extends FilePath {
         try {
             parent = parent.toRealPath();
         } catch (IOException e) {
-            parent = toRealPath(parent);
+            parent = parentToRealPath(parent);
         }
         return parent.resolve(path.getFileName());
     }

